ext/snowcrash/src/UriTemplateParser.cc

```diff
@@ -9,207 +9,251 @@
-                result = true;
-                break;
-            }
-            lastBracket = '}';
+    };
+
+    template <>
+    struct report_action<continuous_dots> {
+
+        template<typename Input>
+        static void apply(const Input& in, state& s)
+        {
+            s.message = std::string{"continuous dots '"} + in.string() + std::string{"'"};
+            s.position = in.position().byte;
+            s.length = in.string().length();
         }
-    }
-    return result;
-}
+    };
 
-static bool PathContainsSquareBrackets(const URITemplate& uriTemplate)
-{
-    return (uriTemplate.find('[') != std::string::npos || uriTemplate.find(']') != std::string::npos);
-}
+    template <>
+    struct report_action<pct_triplet> {
 
-static Expressions GetUriTemplateExpressions(const URITemplate& uriTemplate)
-{
-    Expressions expressions;
-    size_t expressionStartPos = 0;
-    size_t expressionEndPos = 0;
-
-    while (expressionStartPos != std::string::npos && expressionEndPos != std::string::npos
-        && expressionStartPos < uriTemplate.length()) {
-        expressionStartPos = uriTemplate.find("{", expressionStartPos);
-        expressionEndPos = uriTemplate.find("}", expressionStartPos);
-        if (expressionStartPos != std::string::npos && expressionEndPos > expressionStartPos) {
-            expressions.push_back(
-                uriTemplate.substr(expressionStartPos + 1, (expressionEndPos - expressionStartPos) - 1));
+        template<typename Input>
+        static void apply(const Input& in, state& s)
+        {
+            s.message = std::string{"percent encoded character '"} + in.string() + std::string{"'"};
+            s.position = in.position().byte;
+            s.length = in.string().length();
         }
-        expressionStartPos++;
-    }
-    return expressions;
-}
+    };
 
-static ClassifiedExpression ClassifyExpression(const Expression& expression)
-{
+}
 
-    VariableExpression variableExpression(expression);
+struct VariableValidityChecker {
+    ParsedURITemplate& result;
+    const mdp::CharactersRangeSet& sourceBlock;
 
-    if (variableExpression.IsExpressionType()) {
-        return variableExpression;
+    void operator()(const state::variable& s) const {
+        // do nothing - variable is parsed correctly, so there is nothing to report
     }
 
-    QueryStringExpression queryStringExpression(expression);
-    if (queryStringExpression.IsExpressionType()) {
-        return queryStringExpression;
+    void operator()(const state::invalid& s) const {
+
+        tao::pegtl::memory_input<> in(s.content, "");
+        error_locator::state state;
+        tao::pegtl::parse<error_locator::grammar<varspec>, error_locator::report_action>(in, state);
+
+        std::stringstream ss;
+        ss << "URI template variable '" << s.content
+           << "' contains invalid " << state.message
+           <<". Allowed characters for expressions are A-Z a-z 0-9 _ and percent encoded characters";
```

Comment:
Should we perhaps suggest how to percent encode the value? We could URI encode the `state.messge` and suggest that to the user.

"Replacing 'x' with %ab` etc

Reply:
Yes, we can prepare simple parser which will suggest replacement. I send it as one of proposals in internal discussion on slack.

It is question if we want to have it in this PR.

Reply:
We know exact character, so it is very simple task

Reply:
Yeah, I'd suggest we don't solve it here. We can make a follow up change to do this once this is merged so we don't snowball here.